[DataCatalog2.0]: `KedroDataCatalog` with dict interface

ElenaKhaustova opened the pull request:
## Description
Solves https://github.com/kedro-org/kedro/issues/4175

## Development notes
See https://github.com/kedro-org/kedro/issues/4175#issuecomment-2400972641


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

noklam commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -112,6 +114,50 @@ def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
+    def keys(self, regex_search: str | None = None) -> List[str]:  # noqa: UP006
+        return self._filter_keys(regex_search)
+
+    def values(self, regex_search: str | None = None) -> List[AbstractDataset]:  # noqa: UP006
```
```suggestion
    def values(self, regex_search: str | None = None) -> list[AbstractDataset]:
```

noklam replied:
You can use `list` directly

ElenaKhaustova replied:
I had to change it to the current method because we have `list()` method defined for catalog and linter thinks you annotate using it rather than built-in `list`

![Screenshot 2024-10-09 at 11 11 43](https://github.com/user-attachments/assets/81c34420-9024-4b12-ab73-d6b2bdfb6d79)

<img width="713" alt="Screenshot 2024-10-09 at 11 14 43" src="https://github.com/user-attachments/assets/94c7f982-db96-4d60-b1e7-c201b5c5af63">

ankatiyar commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -112,6 +114,50 @@ def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
+    def __iter__(self) -> str:
+        yield from self._datasets.keys()
+
+    def __getitem__(self, ds_name: str) -> AbstractDataset | None:
```
I think this should still error `catalog["non_existent"]` while `catalog.get("non_existent")` returns `None`

ElenaKhaustova replied:
Yeah, agree

ankatiyar reviewed (approved):
Played around with the interface and everything works well, just had a small comment about the `__getitem__` method! 💯

Galileo-Galilei commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -112,6 +114,50 @@ def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
+        return self._filter_keys(regex_search)
+
+    def values(self, regex_search: str | None = None) -> List[AbstractDataset]:  # noqa: UP006
+        return [self._datasets[key] for key in self._filter_keys(regex_search)]
```
Just a nitpick, because the syntax is neat,  but to be honest I can't think about any use case that would need to handle values (i.e. datasets objects) without the associated key (i.e. the dataset name). 

Say you have ``catalog.values(regex_search="^country_")`` :  it is very likely that I want to know the exact data (hence its name) I am loading / reassigning / whatever modification I do, and not just iterate over all datasets  matching the criteria. 

A much better filter search would be something specific to the dataset itself (e.g. its type, some metadata it has...), but I definitely understand it is out of scope here. I am not sure adding other filtering is better than just letting people do something like this : 

```python
[ds_name for ds_name in catalog if isinstance(catalog[ds_name], PandasDataset)] 
```

ElenaKhaustova replied:
That totally makes sense.

We have a ticket on improving search options - https://github.com/kedro-org/kedro/issues/3917, so there's a high chance it will be modified in the future anyway. We also discussed whether we should provide any options for filtering since now it will be easy for users to apply it on top of the keys returned as in your example.

Currently, we kept it as part of the previous API - `list()` method. I added it for `values()` to align with `keys()` and `items()`. However, I see it leading to confusion, so I will remove it from the interface for `values()`.

Galileo-Galilei commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -231,30 +273,31 @@ def _get_dataset(
-            return []
-
         try:
             pattern = re.compile(regex_search, flags=re.IGNORECASE)
```
if ``regex_search`` is forced to be a string, should we add ``regex_kwargs`` everywhere to handle these flags?

Galileo-Galilei commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -264,6 +307,11 @@ def list(self, regex_search: str | None = None) -> list[str]:
+        """Save data to a registered dataset."""
+        self.save_data(name, data)
+
+    def save_data(self, name: str, data: Any) -> None:
```
I am not sure of what is the rationale of keeping ``load_data`` and ``save_data``, but I see this was decided in a TD. Could ou juste reexplain the expected benefit? (I suspect that we keep ``load_data`` to deal with version ; and we keep ``save_data`` for "symmetry")

ElenaKhaustova replied:
You're right about `load_data,` and we are keeping both for now to ensure compatibility with the old catalog. Version will be probably moved from the catalog methods interface soon and technically we won't need `load_data` and `save_data` as we can replace them with `catalog["dataset_name"].load()` and `catalog["dataset_name"].save()`.

ElenaKhaustova replied:
We had one more discussion on this and decided that we will probably keep `load` and `save` at the catalog level, so framework components interact with the datasets via catalog abstraction.

Galileo-Galilei replied:
Yes, I was confused by the renaming but it makes more sense like this 👍

Galileo-Galilei commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -112,6 +114,50 @@ def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
+        if isinstance(value, AbstractDataset):
+            self._datasets[key] = value
+        else:
+            self._datasets[key] = MemoryDataset(data=value)  # type: ignore[abstract]
```
I really (really) think we should have at least logger.info() here to explain this behaviour. I expect silent error due to people assigning a malformed dataset manually which end up assigned as a MemoryDataset, opening a room for a lot of headache.

ElenaKhaustova replied:
I don't mind adding it, though previously, we didn't log it in `add_feed_dict()`, and there were no common issues with that. However, it could be because not many people used this method.

Galileo-Galilei replied:
Wow, I did not know it was already possible ; however as you say I think we target notebooks users with this design, so we can expect it to happen more often that it does now.

Galileo-Galilei commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -231,30 +273,31 @@ def _get_dataset(
     def add(
```
Can (should?) we add a method to remove a dataset? Either ``pop()`` or ``del`` (I saw elsewhere you did not ant to maintain all dict like methods, but I think this may be useful). I think it could be useful when one copy a catalog, filter a pipeline and just want to get some datasets associated to the filtered pipeline.

ElenaKhaustova replied:
We didn't plan to add them in the first version, but we can consider doing so if we see user demand. The main reason for not adding them now is that we still don't want to make catalog fully mutable.

Galileo-Galilei replied:
Yes, I understand that we want to avoid users to modify the catalog inplace in a hook. As a user I'd assume I could handle it like a dict, but this deleting a key prgramatically seems a very rare use case.

merelcht reviewed (approved):
This looks very neat! I think after this change it might be a good moment to add some documentation about using `KedroDataCatalog`. Of course, we'll make it clear it's still in "beta", but I think this will help the team itself get familiar with it and it's behaviour.

ElenaKhaustova commented:
After initial feedback on the interface we removed `regex_search` argument from `keys()`, `values()` and `items()` as it differs from the standard `dict` interface and looked not intuitive to users. We added `filter()` method instead for compatibility with the old `list()` method.

idanov commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -231,40 +272,77 @@ def _get_dataset(
 
+        return self.filter(regex_search)
+
+    def filter(
```
Why I am usually in favour of simpler names, here `filter` is a bit ambiguous, since it's unclear what we are filtering exactly. If we are to keep a separate method, I'd rather keep `list()` with the regex.

ElenaKhaustova replied:
Done

Galileo-Galilei replied:
I quite disagree with this one, because "list" already existed in previous version and virtually no one new about it being able to filter with regex as shown in the interview, so I'd argue that's not helping discovery. 

I personally find `filter` more explicit, or even the original `.values(regex=...)`. What do other think?

ElenaKhaustova replied:
@Galileo-Galilei we agreed to rename it based on the solution for https://github.com/kedro-org/kedro/issues/3917, so we don't rename it several times. That's why we kept the old name for now.

idanov commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -231,40 +272,77 @@ def _get_dataset(
 
     def save(self, name: str, data: Any) -> None:
-        """Save data to a registered dataset."""
+        # TODO: rename input argument when breaking change: name -> ds_name
```
We can still keep it a name, it's not too bad - this will avoid the breaking change.

ElenaKhaustova replied:
We agreed to rename it for consistency with arguments for other methods.

idanov commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -277,7 +355,35 @@ def save(self, name: str, data: Any) -> None:
 
     def load(self, name: str, version: str | None = None) -> Any:
-        """Loads a registered dataset."""
+        # TODO: rename input argument when breaking change: name -> ds_name
```
Same as for save, let's keep it `name`.

ElenaKhaustova replied:
We agreed to rename it for consistency with arguments for other methods.

idanov commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -277,7 +355,35 @@ def save(self, name: str, data: Any) -> None:
     def load(self, name: str, version: str | None = None) -> Any:
-        """Loads a registered dataset."""
+        # TODO: rename input argument when breaking change: name -> ds_name
+        # TODO: remove version from input arguments when breaking change
```
👏

idanov commented on kedro/io/kedro_data_catalog.py:
```diff
@@ -112,6 +114,49 @@ def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
         )
 
+    def keys(self) -> List[str]:  # noqa: UP006
+        return list(self.__iter__())
```
```suggestion
        return list(self)
```
Would this work?

ElenaKhaustova replied:
I made it for iteration cases, like

```
    def values(self) -> List[AbstractDataset]:  # noqa: UP006
        return [self._datasets[key] for key in self]
```
